xdsl/transforms/eqsat_add_costs.py

```diff
@@ -0,0 +1,67 @@
+from xdsl.context import MLContext
+from xdsl.dialects import builtin, eqsat
+from xdsl.dialects.builtin import IntAttr
+from xdsl.ir import OpResult
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+class AddCostEclass(RewritePattern):
+    """
+    Add cost to each operator in an e-class.
+    """
+
+    EQSAT_COST_LABEL = "eqsat_cost"
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: eqsat.EClassOp, rewriter: PatternRewriter):
+        for operand in op.operands:
+            if not isinstance(operand, OpResult):
+                # only add costs to operators
+                continue
+
+            # Add cost attribute to operators
+            operation = operand.op
+            operation.attributes[self.EQSAT_COST_LABEL] = IntAttr(1)
```

Comment:
I have a feeling that it would be good to start with a slightly different approach, I propose the following:
1. walk the IR to find all the eclass ops
2. get their parent blocks and store in a `set`
3. on each block call a helper called `add_eclass_costs(block: Block` that walks the ops and looks for a cost function for each op class
4. do not add costs for unknown functions
5. add a dummy cost function for the ops in this example, we can get away with a hardcoded dictionary of cost functions for the foreseeable future, can make it extensible after the end-to-end flow works

What do you think about this? One of the things that worries me about the example in this PR is that the first eclass op is the op of one of the operands, but I don't think we want to add costs to those here.

Reply:
This might be too complicated for the toy example, and could be a separate pass?
I am thinking of the following:
1. A pass to add costs as attributes (cost = 0 for every non-eclass op, basically what this PR does)
2. A pass to update cost values (this is what you proposed, I think)

Reply:
Maybe I misunderstood the code... Yeah, that would be helpful. Thanks!